gk/fib.c

```diff
@@ -1623,3 +1624,286 @@ del_fib_entry(const char *ip_prefix, struct gk_config *gk_conf)
+	case GK_FWD_GRANTOR:
+		ret = fillup_grantor_fib_dump_entry(dentry, fib);
+		break;
+
+	case GK_FWD_GATEWAY_FRONT_NET:
+		/* FALLTHROUGH */
+	case GK_FWD_GATEWAY_BACK_NET:
+		ret = fillup_gateway_fib_dump_entry(dentry, fib);
+		break;
+
+	case GK_FWD_NEIGHBOR_FRONT_NET:
+		/* FALLTHROUGH */
+	case GK_FWD_NEIGHBOR_BACK_NET:
+		ret = fillup_neighbor_fib_dump_entry(dentry, fib);
+		break;
+
+	case GK_DROP:
+		ret = 0;
+		break;
+
+	default:
+		rte_panic("Invalid FIB action (%u) at %s with lcore %u\n",
+			fib->action, __func__, rte_lcore_id());
+		break;
+	}
+
+	return ret;
+}
+
+int
+print_fib_dump_entry(void *fib_dump_entry, __attribute__((unused)) void *arg)
+{
+	int ret;
+	int i = 0;
+	char ip_addr[128];
+	struct gk_fib_dump_entry *dentry = fib_dump_entry;
+
+	ret = convert_ip_to_str(&dentry->addr, ip_addr, sizeof(ip_addr));
+	if (ret < 0)
+		return -1;
+
+	RTE_LOG(INFO, GATEKEEPER,
+		"FIB entry for IP prefix: %s/%d with action %d.\t",
+		ip_addr, dentry->len, dentry->action);
+
+	if (dentry->action != GK_FWD_GRANTOR)
+		goto ether_cache_tbl;
+
+	ret = convert_ip_to_str(&dentry->grantor_ip, ip_addr, sizeof(ip_addr));
+	if (ret < 0)
+		return -1;
+
+	RTE_LOG(INFO, GATEKEEPER,
+		"Grantor IP address: %s.\n", ip_addr);
+
+ether_cache_tbl:
+	RTE_LOG(INFO, GATEKEEPER,
+		"#Ethernet cache entries: %d.\n", dentry->num_ether_entries);
+
+	while (i < dentry->num_ether_entries) {
+		char d_buf[ETHER_ADDR_FMT_SIZE];
+		char s_buf[ETHER_ADDR_FMT_SIZE];
+		struct gk_fib_ether_entry *eth_entry = &dentry->ether_tbl[i];
+		struct ether_hdr *eth_hdr = &eth_entry->l2_hdr.eth_hdr;
+		struct vlan_hdr *vlan_hdr = &eth_entry->l2_hdr.vlan_hdr;
+
+		ret = convert_ip_to_str(&eth_entry->nexthop_ip,
+			ip_addr, sizeof(ip_addr));
+		if (ret < 0)
+			return -1;
+
+		ether_format_addr(d_buf, sizeof(d_buf), &eth_hdr->d_addr);
+		ether_format_addr(s_buf, sizeof(s_buf), &eth_hdr->s_addr);
+
+		RTE_LOG(INFO, GATEKEEPER,
+			"\tEther cache entry#%d: [state: %s, nexthop ip: %s, ref_cnt: %d, ether_type: 0x%x, d_addr: %s, s_addr: %s, vlan_tci: %hu, vlan_eth_proto: %hu].\n",
+			(i + 1), eth_entry->stale ? "stale" : "fresh", ip_addr,
+			eth_entry->ref_cnt, eth_hdr->ether_type, d_buf, s_buf,
+			vlan_hdr->vlan_tci, vlan_hdr->eth_proto);
+
+		i++;
+	}
+
+	return 0;
+}
+
+int
+list_fib_entries(struct gk_config *gk_conf,
+	gk_fib_dump_entry_handler handler, void *arg)
+{
+	int ret;
+	int index;
+	uint8_t ip6[RTE_LPM6_IPV6_ADDR_SIZE];
+	struct gk_fib *fib;
+	struct gk_lpm *ltbl = &gk_conf->lpm_tbl;
+	const struct rte_lpm_rule *re4;
+	const struct rte_lpm6_rule *re6;
+	struct rte_lpm_iterator_state state;
+	struct rte_lpm6_iterator_state state6;
+	struct gk_fib_dump_entry dentry;
+
+	ret = rte_lpm_iterator_state_init(ltbl->lpm, 0, 0, &state);
+	if (ret < 0) {
+		RTE_LOG(ERR, GATEKEEPER,
+			"gk: failed to initialize the lpm rule iterator state at %s!\n",
+			__func__);
+		return -1;
+	}
+
+	memset(ip6, 0, sizeof(ip6));
+	ret = rte_lpm6_iterator_state_init(ltbl->lpm6, ip6, 0, &state6);
```

Comment:
It looks like you can pass in `NULL` instead of `ip6` to mean an all-zero IPv6 address (according to the header file). Therefore, I think we can remove the variable `ip6`.